privaterelay/fxa_utils.py

```diff
@@ -0,0 +1,156 @@
+        "client_id": client_id,
+        "client_secret": client_secret,
+    }
+
+    expires_in = (social_token.expires_at - datetime.now(timezone.utc)).total_seconds()
+    token = {
+        "access_token": social_token.token,
+        "refresh_token": social_token.token_secret,
+        "token_type": "Bearer",
+        "expires_in": expires_in,
+    }
+
+    client = OAuth2Session(
+        client_id,
+        scope=settings.SOCIALACCOUNT_PROVIDERS["fxa"]["SCOPE"],
+        token=token,
+        auto_refresh_url=refresh_token_url,
+        auto_refresh_kwargs=extra,
+        token_updater=_token_updater,
+    )
+    return client
+
+
+def get_subscription_data_from_fxa(
+    social_account: SocialAccount, fxa_url: str = ""
+) -> dict[str, Any]:
+    if fxa_url == "":
+        # https://github.com/mozilla/fxa/blob/main/packages/fxa-profile-server/docs/API.md#get-v1subscriptions
+        fxa_url = FirefoxAccountsOAuth2Adapter.profile_url + "/v1/subscriptions"
+    try:
+        client = _get_oauth2_session(social_account)
+    except NoSocialToken as e:
+        sentry_sdk.capture_exception(e)
+        return {}
+
+    try:
+        resp = client.get(fxa_url)
+        json_resp = cast(dict[str, Any], resp.json())
+        if "Requested scopes are not allowed" in json_resp.get("message", ""):
+            social_token = SocialToken.objects.get(account=social_account)
+            # refresh user token to get subscription data
+            new_token = client.refresh_token(
+                FirefoxAccountsOAuth2Adapter.access_token_url
+            )
+            update_social_token(social_token, new_token)
+            json_resp = {"social_token": new_token, "refreshed": True}
+    except CustomOAuth2Error as e:
+        sentry_sdk.capture_exception(e)
+    return json_resp
+
+
+def get_phone_subscription_dates(social_account):
+    date_subscribed_phone = start_date = end_date = None
+
+    detailed_subscription_endpoint = settings.FXA_SUBSCRIPTION_ENDPOINT
+    subscription_data = get_subscription_data_from_fxa(
+        social_account, detailed_subscription_endpoint
+    )
+    if "refreshed" in subscription_data.keys():
+        # user token refreshed for expanded scope
+        social_account.refresh_from_db()
+        subscription_data = get_subscription_data_from_fxa(
+            social_account, detailed_subscription_endpoint
+        )
+    if "subscriptions" not in subscription_data.keys():
+        # failed to get subscriptions data which may mean user never had subscription and/or there is data mismatch with FxA
+        profile = social_account.user.profile
+        profile.date_subscribed_phone = None
+        profile.save()
+        # User who was flagged for having phone subscriptions did not actually have phone subscriptions
+        logger.warning("no_subscription_data")
+        return date_subscribed_phone, start_date, end_date
+
+    product_w_phone_capabilites = [settings.PHONE_PROD_ID, settings.BUNDLE_PROD_ID]
```

Comment:
`relay_phones` from `settings.SUBSCRIPTIONS_WITH_PHONE` is a capability tied to products which does not show up in the `FXA_ACCOUNTS_ENDPOINT` subscription call. The way Relay currently gets the `settings.SUBSCRIPTIONS_WITH_PHONE` in the `profile.fxa.extra_data` is by calling the `update_fxa` which uses the FXA's profile url described in `FirefoxAccountsOAuth2Adapter.profile_url` described [here](https://github.com/pennersr/django-allauth/blob/master/allauth/socialaccount/providers/fxa/views.py#L20) (FXA documentation for the url [here](https://github.com/mozilla/fxa/blob/main/packages/fxa-profile-server/docs/API.md#get-v1profile)). There is no way to then tie the `relay_phones` to what subscription the user subscribed to (roundabout way of saying that `SUBSCRIPTIONS_WITH_PHONE` is not a subscription, it's capability).

Happy to create a new variable in `settings` like `settings.PRODUCTS_WITH_PHONE_CAPABILITIES` so it's uniform if other parts of the code calls this. This should not fail in stage or dev since the capability `relay-phones` is tied to product IDs.

Reply:
Thanks for the explanation. It seems like a potential improvement on the FxA side to allow this to be driven by their data rather than our copy.

Let's stick with whatever will work in prod, and we'll take a future action to document how all the Stripe / FxA Subscription stuff works across prod, stage, dev, and local dev.